Code/Tools/RemoteConsole/Core/RemoteConsoleCore.cpp

```diff
@@ -93,40 +93,63 @@ bool RCON_IsRemoteAllowedToConnect(const AZ::AzSock::AzSocketAddress& connectee)
 /////////////////////////////////////////////////////////////////////////////////////////////
 /////////////////////////////////////////////////////////////////////////////////////////////
 /////////////////////////////////////////////////////////////////////////////////////////////
+void SRemoteThreadedObject::Start(const char* name)
+{
+    AZStd::thread_desc desc;
+    desc.m_name = name;
+
+    auto function = AZStd::bind(&SRemoteThreadedObject::ThreadFunction, this);
+    m_thread = AZStd::thread(function, &desc);
+}
+
+void SRemoteThreadedObject::WaitForThread()
+{
+    if (m_thread.joinable())
+    {
+        m_thread.join();
+    }
+}
+
+void SRemoteThreadedObject::ThreadFunction()
+{
+    Run();
+    Terminate();
+}
 
+/////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////
 void SRemoteServer::StartServer()
 {
     StopServer();
     m_bAcceptClients = true;
-    Start(0, kServerThreadName);
+    Start(kServerThreadName);
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////////
 void SRemoteServer::StopServer()
 {
-    Stop();
     m_bAcceptClients = false;
     AZ::AzSock::CloseSocket(m_socket);
     m_socket = SOCKET_ERROR;
-    m_lock.Lock();
-    for (TClients::iterator it = m_clients.begin(); it != m_clients.end(); ++it)
     {
-        it->pClient->StopClient();
+        AZStd::lock_guard<AZStd::recursive_mutex> lock(m_mutex);
+        for (TClients::iterator it = m_clients.begin(); it != m_clients.end(); ++it)
+        {
+            it->pClient->StopClient();
+        }
     }
-    m_lock.Unlock();
-    m_stopEvent.Wait();
-    m_stopEvent.Set();
-}
+    AZStd::unique_lock<AZStd::recursive_mutex> lock(m_mutex);
+    m_stopCondition.wait(lock, [this] { return m_clients.empty(); });}
```

Comment:
the '}' should be on the next line, and do you want to call m_stopCondition.notify_all() here?

Reply:
Actually the condition variable requires a lock in order for it to function.

block scope above should be removed and there should only be one call to lock the mutex.
The condition variable will release the mutex when it puts the thread to sleep